cli/download_source.go

```diff
@@ -0,0 +1,259 @@
+// code is downloaded (or rather, copied) every single time. See the processTerraformSource method for more info.
+func alreadyHaveLatestCode(terraformSource *TerraformSource) (bool, error) {
+	if 	isLocalSource(terraformSource.CanonicalSourceURL) ||
+		!util.FileExists(terraformSource.DownloadDir) ||
+		!util.FileExists(terraformSource.VersionFile) {
+
+		return false, nil
+	}
+
+	currentVersion := encodeSourceVersion(terraformSource.CanonicalSourceURL)
+	previousVersion, err := readVersionFile(terraformSource)
+
+	if err != nil {
+		return false, err
+	}
+
+	return previousVersion == currentVersion, nil
+}
+
+// Return the version number stored in the DownloadDir. This version number can be used to check if the Terraform code
+// that has already been downloaded is the same as the version the user is currently requesting. The version number is
+// calculated using the encodeSourceVersion method.
+func readVersionFile(terraformSource *TerraformSource) (string, error) {
+	return util.ReadFileAsString(terraformSource.VersionFile)
+}
+
+// Write a file into the DownloadDir that contains the version number of this source code. The version number is
+// calculated using the encodeSourceVersion method.
+func writeVersionFile(terraformSource *TerraformSource) error {
+	version := encodeSourceVersion(terraformSource.CanonicalSourceURL)
+	return errors.WithStackTrace(ioutil.WriteFile(terraformSource.VersionFile, []byte(version), 0640))
+}
+
+// Take the given source path and create a TerraformSource struct from it, including the folder where the source should
+// be downloaded to. Our goal is to reuse the download folder for the same source URL between Terragrunt runs.
+// Otherwise, for every Terragrunt command, you'd have to wait for Terragrunt to download your Terraform code, download
+// that code's dependencies (terraform get), and configure remote state (terraform remote config), which is very slow.
+// 
+// To maximize reuse, given a working directory w and a source URL s, we download the code into the folder /T/W/S where:
+//
+// 1. T is the OS temp dir (e.g. /tmp).
+// 2. W is the base 64 encoded sha1 hash of w. This ensures that if you are running Terragrunt concurrently in
+//    multiple folders (e.g. during automated tests), then even if those folders are using the same source URL s, they
+//    do not overwrite each other.
+// 3. S is the base 64 encoded sha1 has of s without its query string. For remote source URLs (e.g. Git
+//    URLs), this is based on the assumption that the scheme/host/path of the URL 
+//    (e.g. git::github.com/foo/bar//some-module) identifies the module name, and we always want to download the same
+//    module name into the same folder (see the encodeSourceName method). We also assume the version of the module is
+//    stored in the query string (e.g. ref=v0.0.3), so we store the base 64 encoded sha1 of the query string in a
+//    file called .terragrunt-source-version within S.
```

Comment:
s/S is the base 64 encoded sha1 **has** of s/S is the base 64 encoded sha1 of s/